Update _iStatFlag for stash items

obligaron opened the pull request:
Fixes a issue mentioned in stash pr #3853

> If a player doesn't have the requirements of an item it is placed on the stash tinted as red, if another player opens the stash it will be still red. The opposite is also true, items show colored even if the player doesn't have the requirements

This pr updates `_iStatFlag` for the active page when opening the stash or changing the active page.
This pr also reduces the number of places where max stash pages are defined.

qndel commented on Source/qol/stash.cpp:
```diff
@@ -538,6 +537,21 @@ void StashStruct::RemoveStashItem(uint16_t iv)
 
+void StashStruct::SetPage(int newPage)
+{
+	page = clamp(newPage, 0, 49);
```
I think it'd be good to extract the 49 to a separate variable for visibility and change it to 99 :)

obligaron replied:
I generalized `CountStashPages` and used it here and for `AutoPlaceItemInStash`.
I would like to separate the change to 100 pages from this bugfix-pr.
But I'm not against more pages. 😁

ephphatha commented on Source/stores.cpp:
```diff
@@ -2169,6 +2169,8 @@ void BarmaidEnter()
 		stextflag = STORE_NONE;
 		IsStashOpen = true;
+		// Forces the page to be refreshed (are items valid to use ...)
+		Stash.SetPage(Stash.GetPage());
```
Having a `Stash::RefreshItemStatFlags()` (or similar) method would be good here instead of a weird looking call to set the page to the current page :D

obligaron replied:
Thanks. I added the method. 🙂

obligaron commented:
With the last change I changed that all items are updated, not only the visible one.
This should help for a text based stash version.

Another thing I changed is that now `_iStatFlag` is also updated when the player changes gears. This was missing in the initial version.

ephphatha commented on Source/qol/stash.cpp:
```diff
@@ -542,6 +543,19 @@ void StashStruct::RemoveStashItem(uint16_t iv)
+
+void StashStruct::RefreshItemStatFlags()
+{
+	// Iterate through active page and update if a item is usable
```
Was tempted to suggest changing this loop but you were a step ahead :D. Only thing I'd be hesitant about is if potential slowdown when opening the stash if someone ends up with more than a few hundred items in the stash. It's a pretty fast operation anyway so I don't think this'll be noticeable even with a full stash.

```suggestion
```
The doxygen comment on the method is more accurate than this inline comment, no value keeping this one.

obligaron replied:
Gone 😉

AJenbo reviewed (approved):
Much better now that it's not part of SetPage :+1:

glebm commented on Source/qol/stash.cpp:
```diff
@@ -423,7 +424,7 @@ uint16_t CheckStashHLight(Point mousePosition)
 	ClearPanel();
 
-	uint16_t itemId = abs(Stash.stashGrids[Stash.page][slot.x][slot.y]);
+	uint16_t itemId = abs(Stash.stashGrids[Stash.GetPage()][slot.x][slot.y]);
```
This calls `abs` on a `uint16_t`. Redundant or a bug?

AJenbo replied:
Redundant, leftover from back when they where `int8_t`

AJenbo replied:
Will be fixed once this is merged: https://github.com/diasurgical/devilutionX/pull/4239/files#diff-fcb318e528e35c32e1bb412094800c059f6bea7a86003f15541e06bde8116fc0R415
